src/audit/routes/query.py

```diff
@@ -0,0 +1,276 @@
+
+        if key in {"start", "stop", "count"}:
+            continue
+
+        if key == "groupby":
+            groupby.add(value)
+            field = value
+        else:
+            query_params[key].add(value)
+            field = key
+
+        try:
+            getattr(model, field)
+        except AttributeError as e:
+            raise HTTPException(
+                HTTP_400_BAD_REQUEST,
+                f"'{field}' is not allowed on category '{category}'",
+            )
+
+    if not config["QUERY_USERNAMES"] and (
+        "username" in query_params or "username" in groupby
+    ):
+        raise HTTPException(
+            HTTP_400_BAD_REQUEST,
+            f"Querying by username is not allowed",
+        )
+
+    if groupby:
+        logs = await query_logs_groupby(
+            model, start_date, stop_date, query_params, groupby
+        )
+        next_timestamp = None
+    else:
+        logs, next_timestamp = await _query_logs(
+            model, start_date, stop_date, query_params, count
+        )
+
+    if not config["QUERY_USERNAMES"]:
+        # TODO: excluding usernames from the query might be more efficient
+        for log in logs:
+            if "username" in log:
+                del log["username"]
+
+    return {
+        "nextTimeStamp": next_timestamp,
+        "data": len(logs) if count else logs,
+    }
+
+
+def add_filters(model, query, query_params, start_date=None, stop_date=None):
+    if start_date:
+        query = query.where(model.timestamp >= start_date)
+    if stop_date:
+        query = query.where(model.timestamp < stop_date)
+    for field, values in query_params.items():
+        # TODO for resource_paths, implement filtering in a way that
+        # would return "/A/B" when querying "/A".
+        if hasattr(getattr(model, field).type, "item_type"):  # ARRAY
+            query = query.where(getattr(model, field).overlap(values))
+        else:
+            query = query.where(db.or_(getattr(model, field) == v for v in values))
+    return query
+
+
+async def _query_logs(model, start_date, stop_date, query_params, count):
+    # get all logs matching the filters and apply the page size limit
+    query = add_filters(model, model.query, query_params, start_date, stop_date)
+    query = query.order_by(model.timestamp)
+    if not count:
+        limit = config["QUERY_PAGE_SIZE"]
+        query = query.limit(limit)
+    logs = await query.order_by(model.timestamp).gino.all()
+
+    if not logs or count:
+        # `count` queries are not paginated: no next timestamp
+        return logs, None
+
+    # if there are more logs with the same timestamp as the last queried
+    # log, also return them
+    last_timestamp = logs[-1].timestamp
+    all_timestamps_query = add_filters(
+        model, model.query, query_params, start_date, stop_date
+    )
+    all_timestamps_query = all_timestamps_query.order_by(model.timestamp)
+    extra_logs = await all_timestamps_query.where(
```

Comment:
I think we'll have to do this as long as we use timestamp as the primary key for our queries and sorting. I think we'd have to add something like `request.uuid` (or whatever FastAPI does) to the records if we cared enough about enforcing page sizes while also allowing for reliable sort order.

But I don't think we need that until we start caring about access patterns/apps that lean on the audit service.